x/sync/network_server_test.go

```diff
@@ -0,0 +1,288 @@
+// Copyright (C) 2019-2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package sync
+
+import (
+	"context"
+	"math/rand"
+	"testing"
+
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/snow/engine/common"
+	"github.com/ava-labs/avalanchego/utils/constants"
+	"github.com/ava-labs/avalanchego/utils/logging"
+	"github.com/ava-labs/avalanchego/x/merkledb"
+	"github.com/golang/mock/gomock"
+
+	"github.com/stretchr/testify/require"
+)
+
+func Test_Server_GetRangeProof(t *testing.T) {
+	r := rand.New(rand.NewSource(1)) // #nosec G404
+
+	smallTrieDB, _, err := generateTrieWithMinKeyLen(t, r, defaultRequestKeyLimit, 1)
+	require.NoError(t, err)
+	smallTrieRoot, err := smallTrieDB.GetMerkleRoot(context.Background())
+	require.NoError(t, err)
+
+	tests := map[string]struct {
+		request             *RangeProofRequest
+		modifyResponse      func(*merkledb.RangeProof)
+		expectedErr         error
+		expectedResponseLen int
+		nodeID              ids.NodeID
+		proofNil            bool
+	}{
+		"proof too small": {
+			request: &RangeProofRequest{
+				Root:       smallTrieRoot,
+				KeyLimit:   defaultRequestKeyLimit,
+				BytesLimit: 1000,
+			},
+			proofNil: true,
+		},
+		"byteslimit is 0": {
+			request: &RangeProofRequest{
+				Root:       smallTrieRoot,
+				KeyLimit:   defaultRequestKeyLimit,
+				BytesLimit: 0,
+			},
+			proofNil: true,
+		},
+		"keylimit is 0": {
+			request: &RangeProofRequest{
+				Root:       smallTrieRoot,
+				KeyLimit:   defaultRequestKeyLimit,
+				BytesLimit: 0,
+			},
+			proofNil: true,
+		},
+		"keys out of order": {
+			request: &RangeProofRequest{
+				Root:       smallTrieRoot,
+				KeyLimit:   defaultRequestKeyLimit,
+				BytesLimit: constants.DefaultMaxMessageSize,
+				Start:      []byte{1},
+				End:        []byte{0},
+			},
+			proofNil: true,
+		},
+		"key limit too large": {
+			request: &RangeProofRequest{
+				Root:       smallTrieRoot,
+				KeyLimit:   2 * defaultRequestKeyLimit,
+				BytesLimit: constants.DefaultMaxMessageSize,
+			},
+			expectedResponseLen: defaultRequestKeyLimit,
+		},
+		"bytes limit too large": {
+			request: &RangeProofRequest{
+				Root:       smallTrieRoot,
+				KeyLimit:   defaultRequestKeyLimit,
+				BytesLimit: 2 * constants.DefaultMaxMessageSize,
+			},
+		},
+	}
+
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			require := require.New(t)
+			ctrl := gomock.NewController(t)
+			defer ctrl.Finish()
+			sender := common.NewMockSender(ctrl)
+			var proofResult *merkledb.RangeProof
+			sender.EXPECT().SendAppResponse(
+				gomock.Any(), // ctx
+				gomock.Any(),
+				gomock.Any(), // requestID
+				gomock.Any(), // responseBytes
+			).DoAndReturn(
+				func(_ context.Context, _ ids.NodeID, requestID uint32, responseBytes []byte) error {
+					// deserialize the response so we can modify it if needed.
```

Comment:
should we update this comment? eg, deserialize the response so we can store it in proofResult